### PR TITLE
Break link definitions onto multiple lines when needed

### PR DESCRIPTION
**Pull request by j-f1**
Ref #3481.

**Prettier 1.9.2**
[Playground link](https://prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBtAVgVwM4wLSYBOANgLpIAEAFjDAA7ZID0TcAHgIYC2dxcAdJC4AdKKiLE8AdwCWMKnmxUIhfDDl9y1Wg2atOPPoIhcKwkOph9zo8SWlyFxaAHM8lzZRr1GLdt14BITMQDgpnKBcAGnDXCg8BCgBJGABybApCOA5iYgBPTOzcgoiXfgoACTgs6g4ANzgKKRUAE2x+GyhbUq1vXT8DQOMuJg48UoISUagIeWrFOBcuBBgmPLh8Dhm5wgWlldG8LJz88dc8ADMZPjwobgE6Kjp+Dmw6NjMxCdl5dw04Xo6Xz6AJGISHCYSaazKjzbCLZawNYbPBbGFwhEHMbHYpnSKXa5wW73fiPZ6vd4hZwQADWFDCCQAhJ0QFEQBA6OpoNhkKA6IQZLAAOoyFryZAADgADGyYBwAEYisVUZAAJjZ2EFLj4AEVMLM4MgLjl4bLCBxrlqAMImLgcZAgGZQQ1s+XmsA0jYAZToHDAWuQMEImDgbPQ2DYACF3Z6YF77gAZQWGpDG4imkC+wjwwgOu2EGktCBSKCskDwrgyQPB0MgHBwAAqCp5qZNtcFOZgAAVzUt7a307WsgBHTAyLI9jh9o1ttn8iDwoXmugO-lwHMNMscQiEYtd7cIFuhOoQUVlotgGeDtmvGBXjMQTD0J+q+9wAC+76AA)
```sh
--parser markdown
```

**Input:**
```markdown
[just-url]: https://example.com
[url-with-short-title]: https://example.com "title"
[url-with-long-title]: https://example.com "a long, long title. It's really really long. Here have words."

[long]: https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx 
[long-with-title]: https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx "look a title!"

```

**Output:**
```markdown
[just-url]: https://example.com
[url-with-short-title]: https://example.com "title"
[url-with-long-title]: https://example.com "a long, long title. It's really really long. Here have words."
[long]: https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx
[long-with-title]: https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx "look a title!"

```

**Output after this PR:**
[<sub><sub></sub></sub> Preview playground link](https://deploy-preview-3531--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBtAVgVwM4wLSYBOANgLpIAEAFjDAA7ZID0TcAHgIYC2dxcAdJC4AdKKiLE8AdwCWMKnmxUIhfDDl9y1Wg2atOPPoIhcKwkOph9zo8SWlyFxaAHM8lzZRr1GLdt14BITMQDgpnKBcAGnDXCg8BCgBJGABybApCOA5iYgBPTOzcgoiXfgoACTgs6g4ANzgKKRUAE2x+GyhbUq1vXT8DQOMuJg48UoISUagIeWrFOBcuBBgmPLh8Dhm5wgWlldG8LJz88dc8ADMZPjwobgE6Kjp+Dmw6NjMxCdl5dw04Xo6Xz6AJGISHCYSaazKjzbCLZawNYbPBbGFwhEHMbHYpnSKXa5wW73fiPZ6vd4hZwQADWFDCCQAhJ0QFEQBA6OpoNhkKAOIRCBApAAFfkIHkoeoQGQtVmhXDIC45eFsgBGhA4YBpGwAynRNTJIsgYIRMHA2S0IGBFcrzSBDfDVMKNUsODbiCqQOhsGwAEIarW6+4AGUNcHdnogmHo0YATBG7frCI7kCAuPyaZapFA5XRCIaYAB1GXyZAADgADGy8xB4YWNXRU3m4I6GnKsgBHTAyLLOjiuhNs+FcGTG0127CGlx8ACKmFm4aQSo9dpgHFVxZapaQsbZJo41ynAGETOnUzMoOG2Tg4AAVdcS5fwgC+z6AA)
```markdown
[just-url]: https://example.com
[url-with-short-title]: https://example.com "title"
[url-with-long-title]:
  https://example.com
  "a long, long title. It's really really long. Here have words."
[long]:
  https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx
[long-with-title]:
  https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx
  "look a title!"

```

### CODE REVIEW

File: src/printer-markdown.js [hunk 1]
```diff
@@ -292,13 +294,21 @@ function genericPrint(path, options, print) {
...
+        concat([
+          concat(["[", node.identifier, "]:"]),
+          indent(
+            concat(
+              [
+                line,
+                printUrl(node.url),
+                node.title ? line : null,
+                node.title ? printTitle(node.title) : null
+              ].filter(x => x)
```
Comment (ikatyang) [inline src/printer-markdown.js]: I think it'd be better this way as it's more intuitive:

```js
concat([
  line,
  printUrl(node.url),
  node.title === null
    ? ""
    : concat([line, printTitle(node.title)])
])
```

And also we should add a test case for empty title:

```md
[something]: http://prettier.io 
```

Comment (j-f1) [reply src/printer-markdown.js]: It seems like the `""` can be 🔥d when the title is blank. Markdown (at least CommonMark) [doesn’t render links with empty titles differently from links without titles](http://spec.commonmark.org/dingus/?text=%5Bfoo%5D%0A%5Bbar%5D%0A%5Bbaz%5D%0A%5Bquux%5D%0A%0A%5Bfoo%5D%3A%20https%3A%2F%2Fexample.com%20%27%27%0A%5Bbar%5D%3A%20https%3A%2F%2Fexample.com%20%22%22%0A%5Bbaz%5D%3A%20https%3A%2F%2Fexample.com%20()%0A%5Bquux%5D%3A%20https%3A%2F%2Fexample.com%0A).

Comment (ikatyang) [reply src/printer-markdown.js]: It'll change the AST (at least `remark`'s), we shouldn't touch it.

Comment (ikatyang) [reply src/printer-markdown.js]: ...sorry, I thought `node.title` was `""` in that case, it seems they're all parsed as `null`. But results are the same anyway.

File: src/printer-markdown.js [hunk 2]
```diff
@@ -292,13 +294,20 @@ function genericPrint(path, options, print) {
...
-        node.title === null ? "" : ` ${printTitle(node.title)}`
-      ]);
+      return group(
+        concat([
+          concat(["[", node.identifier, "]:"]),
+          indent(
+            concat([
+              line,
+              printUrl(node.url),
+              node.title === null
```
Comment (ikatyang) [inline src/printer-markdown.js]: image

**Comment by azz** [comment]
I worry that this will break some renderers (specifically BitBucket Server). Maybe we could make it respect `proseWrap: "preserve"`?

**Comment by ikatyang** [comment]
from [BitBucket Support - Markdown syntax guide](https://confluence.atlassian.com/bitbucketserver/markdown-syntax-guide-776639995.html):

> Bitbucket Server uses Markdown for formatting text, as specified in **CommonMark** (with a few extensions)

It seems they should render it correctly, but they actually didn't follow the CommonMark Spec. I'm not sure what should we do here since the link definition does not look like a prose.

**Comment by j-f1** [comment]
Rebased to fix the merge conflicts.

**Comment by j-f1** [comment]
👍/👎?

**Comment by j-f1** [comment]
Should we merge this or should I 🗑 it?

**Comment by lipis** [comment]
screen shot 2018-02-19 at 11 01 26

Even GitHub is not rendering them correctly (link is not purple).. I would leave them in a single line.. links are links :)

**Comment by j-f1** [comment]
@lipis GitHub renders them correctly:

[just-url] [url-with-short-title] [url-with-long-title] [long] [long-with-title]

[just-url]: https://example.com
[url-with-short-title]: https://example.com "title"
[url-with-long-title]:
  https://example.com
  "a long, long title. It's really really long. Here have words."
[long]:
  https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx
[long-with-title]:
  https://example.com/a-long-url/another-segment/yet-another-segment/a-really-long-file-name.php.aspx
  "look a title!"

**Comment by lipis** [comment]
the result is correct.. but when they are [fenced](https://github.com/prettier/prettier/pull/3531#issue-159195303)?

Either case I would prefer to not break them or at least respect the `proseWrap` option

**Comment by lipis** [comment]
What's the status with this one?

**Comment by j-f1** [comment]
I’m not sure — should this get merged as-is or should it be modified to respect `proseWrap`?

**Comment by lipis** [comment]
I would respect the `proseWrap`.. WDYT?

**Comment by lipis** [comment]
So shall we add the 1.13 tag for a friendly pressure?

**Comment by j-f1** [comment]
Sure 😀
